src/main/java/eu/europeana/apikey/keycloak/KeycloakUserManager.java

```diff
@@ -0,0 +1,253 @@
+import org.springframework.stereotype.Service;
+import org.springframework.http.*;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * Class for working with Keycloak and it's
+ * <a href="https://www.keycloak.org/docs-api/6.0/rest-api/index.html">Rest Admin API</a>.<br/>
+ * <p>
+ * Note that there are 2 client id's used by keycloak:
+ * <ol>
+ *     <li>ClientId which is the same as an apikey (string)</li>
+ *     <li>id which is an internal id (hash) of the client. This is called "id of client (not clientId)" in the keycloak
+ *  *  documentation and saved as keycloakId in an apikey object.</li>
+ * </ol>
+ *  In other words:
+ *  <code>
+ *    apiKey.getId().equals(keycloakClient.getClientId());
+ *    apiKey.getKeycloakId.equals(keycloakClient.getId());
+ *  </code>
+ */
+@Service
+public class KeycloakUserManager {
+    private static final Logger LOG = LogManager.getLogger(KeycloakUserManager.class);
+
+    private static final String MASTER_REALM = "master";
+    private static final String USER_ENDPOINT = "%s/admin/realms/%s/users/%s";
+    private static final String ERROR_COMMUNICATING_WITH_KEYCLOAK = "Error communicating with Keycloak";
+    private static final String RECEIVED = ": received ";
+
+    private final        ObjectMapper          mapper                            = new ObjectMapper();
+    private final        KeycloakTokenVerifier keycloakTokenVerifier;
+    private final        KeycloakTokenVerifier keycloakMasterTokenVerifier;
+    private final        KeycloakProperties    kcProperties;
+
+    /**
+     * Http client used for communicating with Keycloak where Keycloak admin client is not appropriate
+     */
+    private              CloseableHttpClient   httpClient;
+
+    public KeycloakUserManager(KeycloakProperties kcProperties) {
+        this.kcProperties = kcProperties;
+        this.keycloakTokenVerifier = new KeycloakTokenVerifier(kcProperties.getRealmPublicKey());
+        this.keycloakMasterTokenVerifier = new KeycloakTokenVerifier(kcProperties.getMasterPublicKey());
+    }
+
+    @PostConstruct
+    public void init() {
+        httpClient = HttpClients.createDefault();
+    }
+
+    @PreDestroy
+    public void clean() {
+        try {
+            httpClient.close();
+        } catch (IOException e) {
+            LOG.error("Closing http client failed", e);
+        }
+    }
+
+    KeycloakPrincipal<KeycloakSecurityContext> authenticateAdminUser(String username,
+                                                                     String password,
+                                                                     String clientId,
+                                                                     String grantType) {
+        Keycloak keycloak = KeycloakBuilder.builder()
+                                           .realm(MASTER_REALM)
```

Comment:
Why do we need to use master realm here? I would very much prefer to use the europeana realm instead if possible, so we don't have to provide the application with master realm login data

Reply:
Created a new ticket to investigate if this really necessary https://europeana.atlassian.net/browse/EA-2359